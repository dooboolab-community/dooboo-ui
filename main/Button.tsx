import {
  ActivityIndicator,
  LayoutRectangle,
  Platform,
  TouchableOpacity,
} from 'react-native';
import {DoobooTheme, light, withTheme} from './theme';
import React, {useRef, useState} from 'react';
import type {
  StyleProp,
  TextProps,
  TextStyle,
  TouchableOpacityProps,
  ViewStyle,
} from 'react-native';

import type {FC} from 'react';
import styled from '@emotion/native';
import {useHover} from 'react-native-web-hooks';

type Styles = {
  container?: StyleProp<ViewStyle>;
  text?: StyleProp<TextStyle>;
  disabledButton?: StyleProp<ViewStyle>;
  disabledText?: StyleProp<TextStyle>;
  hovered?: StyleProp<ViewStyle>;
};

const Container = styled.View`
  align-self: stretch;
  padding: 8px 12px;
  background-color: ${({theme}) => theme.primary || light.primary};

  flex-direction: row;
  align-items: center;
  justify-content: center;
`;

const Text = styled.Text`
  font-size: 14px;
  color: white;
`;

export interface ButtonProps {
  testID?: string;
  indicatorColor?: string;
  loading?: boolean;
  disabled?: boolean;
  style?: StyleProp<ViewStyle>;
  styles?: Styles;
  leftElement?: React.ReactElement;
  rightElement?: React.ReactElement;
  activeOpacity?: TouchableOpacityProps['activeOpacity'];
  text?: string;
  onPress?: TouchableOpacityProps['onPress'];
  touchableOpacityProps?: Partial<TouchableOpacityProps>;
  textProps?: Partial<TextProps>;
}

const StyledButton: FC<ButtonProps & {theme: DoobooTheme}> = ({
  testID,
  theme,
  disabled,
  loading,
  style,
  styles,
  indicatorColor = theme.disabled,
  leftElement,
  rightElement,
  activeOpacity = 0.7,
  text,
  onPress,
  touchableOpacityProps,
  textProps,
}) => {
  const ref = useRef<TouchableOpacity>(null);
  const hovered = useHover(ref);
  const [layout, setLayout] = useState<LayoutRectangle>();

  const compositeStyles: Styles = {
    disabledButton: {
      backgroundColor: theme.primary,
      borderColor: theme.primary,
    },
    disabledText: {
      color: theme.primary,
    },
    hovered: {
      borderColor: theme.primary,
      shadowColor: 'black',
      shadowOffset: {
        width: 0,
        height: 4,
      },
      shadowOpacity: 0.24,
      shadowRadius: 16,
      elevation: 10,
    },
    ...styles,
  };

  return (
    <TouchableOpacity
      testID={testID}
      ref={Platform.select({
        web: ref,
        default: undefined,
      })}
      activeOpacity={activeOpacity}
      onPress={onPress}
      delayPressIn={50}
      disabled={disabled}
      style={style}
      {...touchableOpacityProps}>
      {loading ? (
        <Container
          testID="loading-view"
          style={[
            compositeStyles.container,
            {
              width: layout?.width,
              height: layout?.height,
            },
            hovered && !disabled && compositeStyles.hovered,
            disabled && compositeStyles.disabledButton,
          ]}>
          <ActivityIndicator size="small" color={indicatorColor} />
        </Container>
      ) : (
        <Container
          testID="button-view"
          style={[
            compositeStyles.container,
            hovered && !disabled && compositeStyles.hovered,
            disabled && compositeStyles.disabledButton,
          ]}
          onLayout={(e) => setLayout(e.nativeEvent.layout)}>
          {leftElement}
          <Text
            style={[
              compositeStyles.text,
              disabled && compositeStyles.disabledText,
            ]}
            {...textProps}>
            {text}
          </Text>
          {rightElement}
        </Container>
      )}
    </TouchableOpacity>
  );
};

StyledButton.defaultProps = {theme: light};

export const Button = withTheme(StyledButton);
