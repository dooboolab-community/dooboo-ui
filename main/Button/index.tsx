import {
  ActivityIndicator,
  LayoutRectangle,
  Platform,
  TouchableOpacity,
} from 'react-native';
import {
  ButtonText,
  ButtonType,
  ButtonWrapper,
} from '../Styled/StyledComponents';
import {DoobooTheme, light, withTheme} from '../theme';
import React, {useRef, useState} from 'react';
import type {
  StyleProp,
  TextProps,
  TextStyle,
  TouchableOpacityProps,
  ViewStyle,
} from 'react-native';
import styled, {css} from '@emotion/native';

import type {FC} from 'react';
import {useHover} from 'react-native-web-hooks';

type Styles = {
  container?: StyleProp<ViewStyle>;
  text?: StyleProp<TextStyle>;
  disabledButton?: StyleProp<ViewStyle>;
  disabledText?: StyleProp<TextStyle>;
  hovered?: StyleProp<ViewStyle>;
};

type ButtonSize = 'small' | 'medium' | 'large';

const ButtonContainer = styled(ButtonWrapper)<{
  type: ButtonType;
  size?: ButtonSize;
  outlined?: boolean;
  disabled?: boolean;
}>`
  align-self: stretch;
  padding: ${({size}) =>
    size === 'large'
      ? '14px 24px'
      : size === 'small'
      ? '4px 12px'
      : '10px 20px'};
  border-radius: ${({size}) => (size === 'large' ? '24px' : '20px')};

  flex-direction: row;
  align-items: center;
  justify-content: center;
`;

export interface ButtonProps {
  testID?: string;
  indicatorColor?: string;
  loading?: boolean;
  disabled?: boolean;
  outlined?: boolean;
  style?: StyleProp<ViewStyle>;
  styles?: Styles;
  leftElement?: React.ReactElement;
  rightElement?: React.ReactElement;
  activeOpacity?: TouchableOpacityProps['activeOpacity'];
  text?: string;
  onPress?: TouchableOpacityProps['onPress'];
  touchableOpacityProps?: Partial<TouchableOpacityProps>;
  textProps?: Partial<TextProps>;
  type?: ButtonType;
  size?: ButtonSize;
}

const ButtonComponent: FC<ButtonProps & {theme: DoobooTheme}> = ({
  testID,
  theme,
  disabled,
  loading,
  style,
  styles,
  indicatorColor = theme.text,
  leftElement,
  rightElement,
  activeOpacity = 0.6,
  text,
  onPress,
  touchableOpacityProps,
  textProps,
  type = 'primary',
  size,
  outlined = false,
}) => {
  const ref = useRef<TouchableOpacity>(null);
  const hovered = useHover(ref);
  const [layout, setLayout] = useState<LayoutRectangle>();

  const compositeStyles: Styles = {
    disabledButton: css`
      background-color: ${!outlined ? theme.disabled : theme.background};
      border-color: ${theme.disabled};
    `,
    disabledText: {
      color: !outlined ? theme.background : theme.textDisabled,
    },
    hovered: {
      shadowColor: 'black',
      shadowOffset: {
        width: 0,
        height: 4,
      },
      shadowOpacity: 0.24,
      shadowRadius: 16,
      elevation: 10,
    },
    ...styles,
  };

  return (
    <TouchableOpacity
      testID={testID}
      ref={Platform.select({
        web: ref,
        default: undefined,
      })}
      activeOpacity={activeOpacity}
      onPress={onPress}
      delayPressIn={50}
      disabled={disabled || loading}
      style={style}
      {...touchableOpacityProps}>
      {loading ? (
        <ButtonContainer
          testID="loading-view"
          disabled
          style={[
            compositeStyles.container,
            {
              width: layout?.width,
              height: layout?.height,
            },
            hovered && !disabled && compositeStyles.hovered,
            disabled && compositeStyles.disabledButton,
          ]}
          type={type}
          size={size}
          outlined={outlined}>
          <ActivityIndicator size="small" color={indicatorColor} />
        </ButtonContainer>
      ) : (
        <ButtonContainer
          testID="button-container"
          style={[
            compositeStyles.container,
            hovered && !disabled && compositeStyles.hovered,
            disabled && compositeStyles.disabledButton,
          ]}
          onLayout={(e) => setLayout(e.nativeEvent.layout)}
          type={type}
          size={size}
          disabled={disabled}
          outlined={outlined}>
          {leftElement}
          <ButtonText
            outlined={outlined}
            type={type}
            style={[
              compositeStyles.text,
              disabled && compositeStyles.disabledText,
            ]}
            {...textProps}>
            {text}
          </ButtonText>
          {rightElement}
        </ButtonContainer>
      )}
    </TouchableOpacity>
  );
};

ButtonComponent.defaultProps = {theme: light};

export const Button = withTheme(ButtonComponent);
