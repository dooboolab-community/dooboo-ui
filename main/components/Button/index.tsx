import type {FC, ReactElement, ReactNode} from 'react';
import {Platform, Text, TouchableOpacity} from 'react-native';
import React, {useCallback, useRef} from 'react';
import type {
  StyleProp,
  TextStyle,
  TouchableOpacityProps,
  ViewStyle,
} from 'react-native';
import styled, {css} from '@emotion/native';

import type {DoobooTheme} from '@dooboo-ui/theme';
import {LoadingIndicator} from '../LoadingIndicator';
import {getTheme} from '../../utils';
import {useHover} from 'react-native-web-hooks';
import {useTheme} from '@dooboo-ui/theme';

type Styles = {
  container?: StyleProp<ViewStyle>;
  text?: StyleProp<TextStyle>;
  disabled?: StyleProp<ViewStyle>;
  disabledText?: StyleProp<TextStyle>;
  hovered?: StyleProp<ViewStyle>;
};

export type ButtonType = 'text' | 'solid' | 'outlined';

export type ButtonColorType =
  | 'primary'
  | 'secondary'
  | 'success'
  | 'danger'
  | 'warning'
  | 'info'
  | 'light';

export type ButtonSizeType = 'small' | 'medium' | 'large';

const ButtonStyles = ({
  theme,
  type = 'solid',
  color = 'primary',
  size = 'medium',
  loading,
  disabled,
}: {
  theme?: DoobooTheme;
  type?: ButtonType;
  color?: ButtonColorType;
  size?: ButtonSizeType;
  disabled?: boolean;
  loading?: boolean;
}): {
  padding?: string;
  backgroundColor?: string;
  borderColor?: string;
  borderWidth?: string;
  textColor?: string;
  disabledBackgroundColor: string;
  disabledBorderColor: string;
  disabledTextColor: string;
} => {
  theme = getTheme(theme);

  let backgroundColor = theme.button[color].bg;
  let borderColor = theme.button[color].bg;
  let textColor = theme.button[color].bg;

  if (disabled) {
    backgroundColor = theme.button.disabled.bg;
    borderColor = theme.button.disabled.text;
    textColor = theme.button.disabled.text;
  }

  if (['text', 'outlined'].includes(type)) {
    backgroundColor = theme.bg.basic;
  }

  if (type === 'solid' || color === 'light') {
    textColor = theme.button[color].text;
  }

  return {
    padding:
      size === 'large'
        ? '16px 32px'
        : size === 'small'
        ? '8px 16px'
        : '12px 24px',
    backgroundColor,
    borderWidth: type === 'outlined' ? '1px' : undefined,
    borderColor,
    textColor,
    disabledBackgroundColor:
      type === 'solid' && !loading ? theme.button.disabled.bg : theme.bg.basic,
    disabledBorderColor: theme.bg.disabled,
    disabledTextColor: theme.button.disabled.text,
  };
};

const ButtonContainer = styled.View<{
  type: ButtonType;
  size?: ButtonSizeType;
  outlined?: boolean;
  disabled?: boolean;
  loading?: boolean;
}>`
  align-self: stretch;
  flex-direction: row;
  align-items: center;
  justify-content: center;
`;

export type Props = {
  testID?: string;
  type?: ButtonType;
  color?: ButtonColorType;
  size?: ButtonSizeType;
  disabled?: boolean;
  loading?: boolean;
  loadingElement?: ReactElement;
  text?: string;
  borderRadius?: number;
  startElement?: ReactElement;
  endElement?: ReactElement;
  style?: StyleProp<ViewStyle>;
  styles?: Styles;
  onPress?: TouchableOpacityProps['onPress'];
  activeOpacity?: TouchableOpacityProps['activeOpacity'];
  touchableOpacityProps?: Partial<TouchableOpacityProps>;
};

export const Button: FC<Props> = (props) => {
  const {
    testID,
    type = 'solid',
    color = 'primary',
    size = 'medium',
    disabled,
    loading = false,
    loadingElement,
    text,
    startElement,
    endElement,
    style,
    styles,
    onPress,
    activeOpacity = 0.6,
    touchableOpacityProps,
    borderRadius,
  } = props;

  const ref = useRef<TouchableOpacity>(null);
  const hovered = useHover(ref);

  const {theme} = useTheme();

  const {
    padding,
    backgroundColor,
    borderColor,
    borderWidth,
    textColor,
    disabledBackgroundColor,
    disabledBorderColor,
    disabledTextColor,
  } = ButtonStyles({theme, type, color, size, loading, disabled});

  const compositeStyles: Styles = {
    container: css`
      padding: ${padding};
      background-color: ${backgroundColor};
      border-color: ${borderColor};
      border-width: ${borderWidth};
    `,
    text: {
      color: textColor,
    },
    disabled: css`
      background-color: ${disabledBackgroundColor};
      border-color: ${disabledBorderColor};
    `,
    disabledText: {
      color: disabledTextColor,
    },
    hovered: {
      shadowColor: 'black',
      shadowOffset: {
        width: 0,
        height: 4,
      },
      shadowOpacity: 0.24,
      shadowRadius: 16,
      elevation: 10,
    },
    ...styles,
  };

  const renderContainer = useCallback(
    (children: ReactNode): ReactElement => {
      return (
        <ButtonContainer
          testID={loading ? 'loading-view' : 'button-container'}
          style={[
            compositeStyles.container,
            hovered && !disabled && compositeStyles.hovered,
            disabled && compositeStyles.disabled,
            {borderRadius: borderRadius ?? 4},
          ]}
          type={type}
          size={size}
          disabled={disabled}
        >
          {children}
        </ButtonContainer>
      );
    },
    [
      borderRadius,
      compositeStyles.container,
      compositeStyles.disabled,
      compositeStyles.hovered,
      disabled,
      hovered,
      loading,
      size,
      type,
    ],
  );

  const LoadingView = loadingElement ?? (
    <LoadingIndicator size="small" color={theme.text.basic} />
  );

  const ChildView = (
    <>
      {startElement}
      <Text
        style={[compositeStyles.text, disabled && compositeStyles.disabledText]}
      >
        {text}
      </Text>
      {endElement}
    </>
  );

  return (
    <TouchableOpacity
      testID={testID}
      ref={Platform.select({
        web: ref,
        default: undefined,
      })}
      activeOpacity={activeOpacity}
      onPress={onPress}
      delayPressIn={50}
      disabled={disabled || loading}
      style={style}
      {...touchableOpacityProps}
    >
      {renderContainer(loading ? LoadingView : ChildView)}
    </TouchableOpacity>
  );
};
