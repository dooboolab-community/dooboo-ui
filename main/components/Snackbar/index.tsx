import {Animated, Dimensions, Platform, TouchableOpacity} from 'react-native';
import {ButtonText, SnackbarWrapper} from '../Styled/StyledComponents';
import type {StyleProp, TextStyle, ViewStyle} from 'react-native';
import {
  forwardRef,
  useCallback,
  useEffect,
  useImperativeHandle,
  useState,
} from 'react';

import type {ReactElement} from 'react';
import type {SnackbarType} from '../Styled/StyledComponents';
import styled from '@emotion/native';
import {useTheme} from '@dooboo-ui/theme';

type Styles = {
  container?: StyleProp<ViewStyle>;
  text?: StyleProp<TextStyle>;
  type?: SnackbarType;
};

export interface SnackbarProps {
  testID?: string;
  ref: React.MutableRefObject<SnackbarRef>;
}

export type SnackbarContent = {
  timer?: SnackbarTimer;
  styles?: Styles;
  actionText?: string;
  onActionPress?: () => void;
  type?: SnackbarType;
  content: {element?: ReactElement; text?: string};
};

interface OpenState {
  isVisible?: boolean;
  isOpen?: boolean;
  timeout?: any;
}

export interface SnackbarRef {
  show(content: SnackbarContent): void;
}

export enum SnackbarTimer {
  SHORT = 1500,
  LONG = 3000,
}

const TextAction = styled.Text<{type: SnackbarType}>`
  color: ${({theme, type}) =>
    type === 'danger' ? theme.text.contrast : theme.text.basic};
`;

const Divider = styled.View<{type: SnackbarType}>`
  width: 1px;
  height: 100%;
  margin: 0 16px;
  opacity: 0.4;

  background-color: ${({theme, type}) =>
    type === 'danger' ? theme.text.contrast : theme.text.basic};
`;

const SnackbarContainer = (
  props: SnackbarProps,
  ref: React.Ref<SnackbarRef>,
): ReactElement | null => {
  const {testID} = props;

  const [openState, setOpened] = useState<OpenState>({
    isVisible: false,
    isOpen: false,
  });

  const [snackbar, setSnackbar] = useState<SnackbarContent>({
    content: {text: ''},
    timer: SnackbarTimer.SHORT,
  });

  const {
    content,
    type = 'primary',
    actionText,
    onActionPress,
    styles,
    timer = SnackbarTimer.SHORT,
  } = snackbar;

  const {theme} = useTheme();

  const {isOpen, isVisible, timeout} = openState;
  const [fadeAnim] = useState(new Animated.Value(0));

  const show = (c: SnackbarContent): void => {
    setSnackbar(c);
    timeout && clearTimeout(timeout);

    setOpened((prevState) =>
      Object.assign(Object.assign({}, prevState), {isOpen: true}),
    );
  };

  const hide = useCallback(
    (duration = 200): void => {
      Animated.timing(fadeAnim, {
        toValue: 0,
        duration,
        useNativeDriver: Platform.select({
          web: false,
          default: true,
        }),
      }).start(() =>
        setOpened((prevState) =>
          Object.assign(Object.assign({}, prevState), {isVisible: false}),
        ),
      );
    },
    [fadeAnim],
  );

  useEffect(() => {
    if (isOpen) {
      if (isVisible) {
        hide(50);
      } else {
        const hideTimeout = setTimeout(() => {
          hide();
        }, timer + 200);

        setOpened({
          isOpen: false,
          isVisible: true,
          timeout: hideTimeout,
        });

        Animated.timing(fadeAnim, {
          toValue: 1,
          duration: 200,
          useNativeDriver: Platform.select({
            web: false,
            default: true,
          }),
        }).start();
      }
    }
  }, [openState, fadeAnim, hide, isOpen, isVisible, timer]);

  useImperativeHandle(ref, () => ({show}));

  return openState.isVisible ? (
    <SnackbarWrapper
      testID={testID}
      type={type}
      style={[
        styles?.container,
        {
          shadowColor: theme.text.basic,
          shadowOffset: {
            width: 0,
            height: 4,
          },
          shadowOpacity: 0.15,
          shadowRadius: 12,
        },
        {maxWidth: Dimensions.get('screen').width - 40},
        {opacity: fadeAnim},
      ]}
    >
      {content.element ? (
        content.element
      ) : (
        <ButtonText style={styles?.text} type={type}>
          {content.text}
        </ButtonText>
      )}
      {actionText ? (
        <Divider
          theme={theme}
          type={type}
          style={{height: Platform.select({web: 24, default: 12})}}
        />
      ) : null}
      {actionText ? (
        <TouchableOpacity onPress={onActionPress}>
          <TextAction theme={theme} type={type} style={styles?.text}>
            {actionText}
          </TextAction>
        </TouchableOpacity>
      ) : null}
    </SnackbarWrapper>
  ) : null;
};

export const Snackbar =
  forwardRef<SnackbarRef, SnackbarProps>(SnackbarContainer);
