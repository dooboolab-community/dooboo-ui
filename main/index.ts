export * from './uis/Typography';
export * from './uis/RadioGroup';
export * from './modals/Snackbar';
export * from './uis/Icon';
export * from './uis/Hr';
export * from './uis/Button';
export * from './uis/ButtonGroup';
export * from './uis/IconButton';
export * from './uis/Checkbox';
export * from './uis/EditText';
export * from './uis/NetworkImage';
export * from './uis/ProgressBar';
export * from './uis/ProgressCircle';
export * from './uis/LoadingIndicator';
export * from './uis/Accordion';
export * from './uis/StatusbarBrightness';
export * from './uis/SwitchToggle';
export * from './uis/Fabs';
export * from './uis/Rating';
export * from './uis/CustomPressable';
export * from './providers';
