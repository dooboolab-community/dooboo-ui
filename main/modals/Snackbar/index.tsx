import type {ReactElement} from 'react';
import React, {forwardRef, useImperativeHandle, useState} from 'react';
import type {StyleProp, TextStyle, ViewStyle} from 'react-native';
import {Modal, Platform, StyleSheet} from 'react-native';
import {useTheme} from '@dooboo-ui/theme';
import styled, {css} from '@emotion/native';

import type {ButtonColorType} from '../../uis/Button';
import {Button} from '../../uis/Button';
import {Icon} from '../../uis/Icon';

import {SnackbarTimer} from './const';

const Container = styled.View`
  flex: 1;
  align-self: stretch;

  flex-direction: row;
  justify-content: center;
`;

const SnackbarContainer = styled.SafeAreaView<{color: ButtonColorType}>`
  background-color: ${({theme, color}) => theme.button[color].bg};
  border-radius: 8px;
  margin-bottom: 52px;
  margin-left: 12px;
  margin-right: 12px;
  align-self: flex-end;

  flex-direction: row;
  align-items: center;
`;

const ActionContainer = styled.View`
  margin-right: 4px;
`;

const SnackbarText = styled.Text<{color: ButtonColorType}>`
  font-family: Pretendard;
  color: ${({theme, color}) => theme.button[color].text};
  flex: 1;
  padding: 12px;
`;

export type SnackbarProps = {
  testID?: string;
  style?: StyleProp<ViewStyle>;
  ref: React.MutableRefObject<SnackbarContext>;
};

export type SnackbarStyles = {
  container?: StyleProp<ViewStyle>;
  text?: StyleProp<TextStyle>;
  actionContainer?: StyleProp<ViewStyle>;
  actionText?: StyleProp<TextStyle>;
};

export type SnackbarOptions = {
  color?: ButtonColorType;
  styles?: SnackbarStyles;
  text?: string;
  actionText?: string;
  timer?: SnackbarTimer | number;
};

export type SnackbarContext = {
  open(snackbarOptions?: SnackbarOptions): void;
  close(): void;
};

let timer: NodeJS.Timeout | null = null;

function clearTimer(): void {
  if (timer) {
    clearTimeout(timer);
    timer = null;
  }
}

function Snackbar(
  {style}: SnackbarProps,
  ref: React.Ref<SnackbarContext>,
): ReactElement {
  const [options, setOptions] = useState<SnackbarOptions | null>(null);
  const [visible, setVisible] = useState(false);
  const {theme} = useTheme();

  useImperativeHandle(ref, () => ({
    open: (snackbarOptions) => {
      clearTimer();
      setVisible(true);
      if (snackbarOptions) {
        setOptions(snackbarOptions);
      }

      timer = setTimeout(() => {
        setVisible(false);
        clearTimer();
      }, snackbarOptions?.timer ?? SnackbarTimer.SHORT);
    },
    close: () => {
      setVisible(false);
      setOptions(null);
    },
  }));

  const {text, styles, actionText, color = 'primary'} = options ?? {};

  const SnackbarContent = (
    <Container>
      <SnackbarContainer
        color={color}
        style={StyleSheet.flatten([
          Platform.OS !== 'web' && {
            shadowOffset: {width: 0, height: 4},
            shadowColor: theme.text.basic,
          },
          styles?.container,
        ])}
      >
        <SnackbarText
          color={color}
          style={StyleSheet.flatten([
            css`
              font-family: 'Pretendard';
              color: ${theme.button[color].text};
            `,
            styles?.text,
          ])}
        >
          {text}
        </SnackbarText>
        <ActionContainer style={styles?.actionContainer}>
          {actionText ? (
            <Button
              type="text"
              onPress={() => setVisible(false)}
              styles={{
                text: StyleSheet.flatten([
                  css`
                    font-family: 'Pretendard';
                    color: ${theme.button[color].text};
                  `,
                  styles?.actionText,
                ]),
              }}
              text={actionText}
            />
          ) : (
            <Button
              type="text"
              onPress={() => setVisible(false)}
              text={<Icon name="X" color={theme.button[color].text} />}
            />
          )}
        </ActionContainer>
      </SnackbarContainer>
    </Container>
  );

  return (
    <Modal
      animationType="fade"
      transparent={true}
      visible={visible}
      style={[
        css`
          flex: 1;
          align-self: stretch;
        `,
        style,
      ]}
    >
      {SnackbarContent}
    </Modal>
  );
}

export default forwardRef<SnackbarContext, SnackbarProps>(Snackbar);
