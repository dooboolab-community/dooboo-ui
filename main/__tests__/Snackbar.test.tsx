import React, {useCallback, useRef} from 'react';
import {RenderAPI, fireEvent, render} from '@testing-library/react-native';

import {Button} from '../Button';
import {Snackbar} from '../../main';
import {SnackbarRef} from '../Snackbar';
import {SnackbarType} from '../Styled/StyledComponents';
import {View} from 'react-native';
import {createComponent} from '../../test/testUtils';

let component: React.ReactElement;
let testingLib: RenderAPI;

function SnackbarDefault(): React.ReactElement {
  const snackbar = useRef<SnackbarRef>(null);

  const onPress = useCallback(
    (type?: SnackbarType): void => {
      if (snackbar)
        if (type === 'danger')
          return snackbar.current?.show({
            text: 'Lorem ipsum dolor sit amet',
            type,
            actionText: 'Action',
          });

      if (type === 'info')
        return snackbar.current?.show({
          text: 'Lorem ipsum dolor sit amet',
          type,
          actionText: 'Action',
        });

      snackbar.current?.show({
        text: 'Lorem ipsum dolor sit amet',
        type,
      });
    },
    [snackbar],
  );

  return (
    <View
      style={{
        flexDirection: 'column',

        flex: 1,
        alignSelf: 'stretch',
        justifyContent: 'center',
        alignItems: 'center',
      }}>
      <Snackbar testID="snackbar" ref={snackbar} />
      <Snackbar testID="snackbar-action" ref={snackbar} />
      <Button
        testID="button"
        type={'primary'}
        onPress={() => onPress('default')}
        text={'Open Snackbar - default'}
        style={{margin: 20}}
      />
      <Button
        testID="button-danger"
        type={'primary'}
        onPress={() => onPress('danger')}
        text={'Open Snackbar - default'}
        style={{margin: 20}}
      />
      <Button
        testID="button-info"
        type={'primary'}
        onPress={() => onPress('info')}
        text={'Open Snackbar - default'}
        style={{margin: 20}}
      />
    </View>
  );
}

describe('Snackbar', () => {
  it('should render test component', () => {
    component = createComponent(<SnackbarDefault />);

    testingLib = render(component);

    const json = testingLib.toJSON();

    expect(json).toMatchSnapshot();
  });

  it('should open snackbar', async () => {
    // const timingSpy = jest.spyOn(Animated, 'timing');

    component = createComponent(<SnackbarDefault />);

    testingLib = render(component);

    const btn = testingLib.getByTestId('button');

    fireEvent.press(btn);
    testingLib.rerender(component);

    // const snackbar = await waitFor(() => testingLib.getByTestId('snackbar'));

    // expect(snackbar).toBeTruthy();

    // expect(
    //   timingSpy.mock.calls.find(
    //     (predicate) =>
    //       predicate[1].toValue === 0 && predicate[1].duration === 50,
    //   ),
    // ).toBeTruthy();
  });

  it('should open snackbar `danger` with action', async () => {
    component = createComponent(<SnackbarDefault />);

    testingLib = render(component);

    const btn = testingLib.getByTestId('button-danger');

    fireEvent.press(btn);
  });

  it('should open snackbar `info` with action', async () => {
    component = createComponent(<SnackbarDefault />);

    testingLib = render(component);

    const btn = testingLib.getByTestId('button-info');

    fireEvent.press(btn);
  });
});
