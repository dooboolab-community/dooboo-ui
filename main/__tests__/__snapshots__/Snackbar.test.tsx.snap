// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snackbar should render test component 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "stretch",
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "center",
    }
  }
>
  <View
    accessibilityState={
      Object {
        "disabled": false,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "margin": 20,
        "opacity": 1,
      }
    }
    testID="button"
  >
    <View
      onLayout={[Function]}
      outlined={false}
      style={
        Array [
          Object {
            "backgroundColor": "black",
            "borderBottomColor": "black",
            "borderLeftColor": "black",
            "borderRightColor": "black",
            "borderTopColor": "black",
          },
          Array [
            Object {
              "alignItems": "center",
              "alignSelf": "stretch",
              "borderBottomLeftRadius": 20,
              "borderBottomRightRadius": 20,
              "borderTopLeftRadius": 20,
              "borderTopRightRadius": 20,
              "flexDirection": "row",
              "justifyContent": "center",
              "paddingBottom": 10,
              "paddingLeft": 20,
              "paddingRight": 20,
              "paddingTop": 10,
            },
            Array [
              undefined,
              false,
              undefined,
            ],
          ],
        ]
      }
      testID="button-container"
      type="primary"
    >
      <Text
        outlined={false}
        style={
          Array [
            Object {
              "color": "white",
            },
            Array [
              undefined,
              undefined,
            ],
          ]
        }
        type="primary"
      >
        Open Snackbar - default
      </Text>
    </View>
  </View>
  <View
    accessibilityState={
      Object {
        "disabled": false,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "margin": 20,
        "opacity": 1,
      }
    }
    testID="button-danger"
  >
    <View
      onLayout={[Function]}
      outlined={false}
      style={
        Array [
          Object {
            "backgroundColor": "black",
            "borderBottomColor": "black",
            "borderLeftColor": "black",
            "borderRightColor": "black",
            "borderTopColor": "black",
          },
          Array [
            Object {
              "alignItems": "center",
              "alignSelf": "stretch",
              "borderBottomLeftRadius": 20,
              "borderBottomRightRadius": 20,
              "borderTopLeftRadius": 20,
              "borderTopRightRadius": 20,
              "flexDirection": "row",
              "justifyContent": "center",
              "paddingBottom": 10,
              "paddingLeft": 20,
              "paddingRight": 20,
              "paddingTop": 10,
            },
            Array [
              undefined,
              false,
              undefined,
            ],
          ],
        ]
      }
      testID="button-container"
      type="primary"
    >
      <Text
        outlined={false}
        style={
          Array [
            Object {
              "color": "white",
            },
            Array [
              undefined,
              undefined,
            ],
          ]
        }
        type="primary"
      >
        Open Snackbar - default
      </Text>
    </View>
  </View>
  <View
    accessibilityState={
      Object {
        "disabled": false,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "margin": 20,
        "opacity": 1,
      }
    }
    testID="button-info"
  >
    <View
      onLayout={[Function]}
      outlined={false}
      style={
        Array [
          Object {
            "backgroundColor": "black",
            "borderBottomColor": "black",
            "borderLeftColor": "black",
            "borderRightColor": "black",
            "borderTopColor": "black",
          },
          Array [
            Object {
              "alignItems": "center",
              "alignSelf": "stretch",
              "borderBottomLeftRadius": 20,
              "borderBottomRightRadius": 20,
              "borderTopLeftRadius": 20,
              "borderTopRightRadius": 20,
              "flexDirection": "row",
              "justifyContent": "center",
              "paddingBottom": 10,
              "paddingLeft": 20,
              "paddingRight": 20,
              "paddingTop": 10,
            },
            Array [
              undefined,
              false,
              undefined,
            ],
          ],
        ]
      }
      testID="button-container"
      type="primary"
    >
      <Text
        outlined={false}
        style={
          Array [
            Object {
              "color": "white",
            },
            Array [
              undefined,
              undefined,
            ],
          ]
        }
        type="primary"
      >
        Open Snackbar - default
      </Text>
    </View>
  </View>
</View>
`;
